### PR TITLE
Reduce UDP service per-packet allocation size

### PR DESCRIPTION
**Pull request by sparrc**
###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass
- [x] CHANGELOG.md updated

### CODE REVIEW

**Comment by sparrc** [comment]
WIP

This change makes the UDP service only create buffers for each packet that are the size of the incoming packet, rather than creating a maximum packet size buffer for each packet.

This will reduce memory pressure and number of GC cycles, my results sending 100,000 UDP points were:

- udp-payload-size=0 (default):  GCs
- udp-payload-size=1500:  GCs
- udp-payload-size=0 (with change):  GCs
- udp-payload-size=1500 (with change): 112 GCs

**Comment by sparrc** [comment]
cc @joelegasse because you've been working on the influxdb relay, might be relevant to the UDP service there

**Comment by joelegasse** [comment]
I actually did something similar in the relay that entirely eliminates the need for the config option, and instead just has a per-listener 64K buffer. :-)

https://github.com/influxdata/influxdb-relay/blob/master/relay/udp.go#L113

**Comment by sparrc** [comment]
this essentially does that too, I just don't really understand the influxdb config enough to know how to deprecate a config option?

**Comment by sparrc** [comment]
Erp, nevermind, I've figured it out

**Comment by sparrc** [comment]
cc @sebito91, I believe you made this change to have a configurable udp-payload-size in the first place. This change goes a bit further and only allocates a buffer on each loop iteration that is the exact size of the incoming packet.

**Comment by sebito91** [comment]
I'm away from the machine at the moment but the reason we put in the option
was to prevent alloc of the entire 64k buffer each time you called read on
the socket...if you do a very cheesy memalloc in a loop and time it you
will see about 10x speed up for a smaller buffer (e.g. 1600 vs 64k).

I will post the findings when I'm back at the terminal.

On Thursday, April 7, 2016, Cameron Sparr <notifications@github.com> wrote:

> cc @sebito91 <https://github.com/sebito91>, I believe you made this
> change to have a configurable udp-payload-size in the first place. This
> change goes a bit further and only allocates a buffer on each loop
> iteration that is the exact size of the incoming packet.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdata/influxdb/pull/6263#issuecomment-207134826>
>


-- 

--

Sebastian Borza
---------------
PGP: EDC2 BF61 4B91 14F2 AAB4  06C9 3744 7F3F E411 0D3E

**Comment by sebito91** [comment]
And re-reading the thread I think this change would actually be awesome,
that was my intent before but my Golang skill set at the time was wonky :/

:thumbsup: from me!

On Thursday, April 7, 2016, Sebastian Borza <sborza@alumni.princeton.edu>
wrote:

> I'm away from the machine at the moment but the reason we put in the
> option was to prevent alloc of the entire 64k buffer each time you called
> read on the socket...if you do a very cheesy memalloc in a loop and time it
> you will see about 10x speed up for a smaller buffer (e.g. 1600 vs 64k).
>
> I will post the findings when I'm back at the terminal.
>
> On Thursday, April 7, 2016, Cameron Sparr <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> cc @sebito91 <https://github.com/sebito91>, I believe you made this
>> change to have a configurable udp-payload-size in the first place. This
>> change goes a bit further and only allocates a buffer on each loop
>> iteration that is the exact size of the incoming packet.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/influxdata/influxdb/pull/6263#issuecomment-207134826>
>>
>
>
> --
>
> --
>
> Sebastian Borza
> ---------------
> PGP: EDC2 BF61 4B91 14F2 AAB4  06C9 3744 7F3F E411 0D3E
>
>
>

-- 

--

Sebastian Borza
---------------
PGP: EDC2 BF61 4B91 14F2 AAB4  06C9 3744 7F3F E411 0D3E

File: services/udp/service.go [hunk 1]
```diff
@@ -153,15 +156,17 @@ func (s *Service) serve() {
...
 				s.statMap.Add(statReadFail, 1)
 				s.Logger.Printf("Failed to read UDP message: %s", err)
 				continue
 			}
 			s.statMap.Add(statBytesReceived, int64(n))
-			s.parserChan <- buf[:n]
+
+			bufCopy := make([]byte, n)
+			copy(bufCopy, buf[:n])
+			s.parserChan <- bufCopy
```
Comment (sebito91) [inline services/udp/service.go]: IIRC we have to zero out buf for each subsequent ReadFromUDP otherwise we fill the buf?

Comment (sparrc) [inline services/udp/service.go]: each call to ReadFromUDP will write to buf from the beginning, so it won't ever fill

Comment (joelegasse) [inline services/udp/service.go]: This is correct the way it's written. `ReadFromUDP` can't modify our slice, so `buf` will always refer to the full 64K slice, and will get data copied into it starting at index zero, and return `n`, the number of bytes that were copied for a given read.

Comment (sebito91) [inline services/udp/service.go]: I'm not sure why I thought (or remembered) differently, but as long as this checks out and passes tests then I'm fine with it.

**Comment by joelegasse** [comment]
LGTM :+1:

**Pull request merged**